### PR TITLE
email notification Back-end commit

### PR DESCRIPTION
**Pull request by DhaaraniCIT**

### CODE REVIEW

File: apps/workspaces/models.py [hunk 1]
```diff
@@ -65,11 +66,13 @@ class WorkspaceSchedule(models.Model):
     Workspace Schedule
     """
     id = models.AutoField(primary_key=True, help_text='Unique Id to identify a schedule')
-    workspace = models.OneToOneField(Workspace, on_delete=models.PROTECT, help_text='Reference to Workspace model',
-                                     related_name='workspace_schedules')
+    workspace = models.OneToOneField(Workspace, on_delete=models.PROTECT, help_text='Reference to Workspace model')
```
Comment (Sravanksk) [inline apps/workspaces/models.py:69]: why is this changed? Any particular reason?

File: apps/workspaces/tasks.py [hunk 2]
```diff
@@ -1,24 +1,59 @@
...
 from django_q.models import Schedule
 
 from apps.fyle.models import ExpenseGroup
 from apps.fyle.tasks import async_create_expense_groups
 from apps.quickbooks_online.tasks import schedule_bills_creation, schedule_cheques_creation, \
     schedule_journal_entry_creation, schedule_credit_card_purchase_creation, schedule_qbo_expense_creation
 from apps.tasks.models import TaskLog
-from apps.workspaces.models import WorkspaceSchedule, WorkspaceGeneralSettings, LastExportDetail
+from apps.workspaces.models import User, Workspace, WorkspaceSchedule, WorkspaceGeneralSettings, LastExportDetail, QBOCredential
+from fyle_accounting_mappings.models import MappingSetting, ExpenseAttribute
```
Comment (Sravanksk) [inline apps/workspaces/tasks.py:18]: move this import above line 12 and `MappingSetting` is imported but not being used anywhere in the code

File: apps/workspaces/tasks.py [hunk 3]
```diff
@@ -131,3 +166,57 @@ def export_to_qbo(workspace_id, export_mode=None):
...
             )
+
+def run_email_notification(workspace_id):
+
+    ws_schedule, _ = WorkspaceSchedule.objects.get_or_create(
+        workspace_id=workspace_id
+    )
+
+    task_logs = TaskLog.objects.filter(
+        ~Q(type__in=['CREATING_REIMBURSEMENT', 'FETCHING_EXPENSES', 'CREATING_AP_PAYMENT']),
```
Comment (Sravanksk) [inline apps/workspaces/tasks.py:177]: This has to be fixed as per QBO task_logs, please check the code and figure this out

File: apps/workspaces/tasks.py [hunk 4]
```diff
@@ -131,3 +166,57 @@ def export_to_qbo(workspace_id, export_mode=None):
...
+                    'qbo_company': qbo.company_name,
+                    'workspace_id': workspace_id,
+                    'export_time': workspace.last_synced_at.date(),
+                    'year': date.today().year,
+                    'app_url': "{0}/workspaces/{1}/expense_groups".format(settings.FYLE_APP_URL, workspace_id)
+                    }
+                message = render_to_string("mail_template.html", context)
+
+                mail = EmailMessage(
+                    subject="Export To QuickBooks Failed",
```
Comment (Sravanksk) [inline apps/workspaces/tasks.py:211]: Quickbooks -> `QuickBooks Online`

File: apps/workspaces/tasks.py [hunk 5]
```diff
@@ -131,3 +166,57 @@ def export_to_qbo(workspace_id, export_mode=None):
...
+                    body=message,
+                    from_email=settings.EMAIL,
```
Comment (Sravanksk) [inline apps/workspaces/tasks.py:213]: add this env variable from_email (`EMAIL`) to settings.py file

File: apps/workspaces/templates/mail_template.html [hunk 6]
```diff
@@ -0,0 +1,206 @@
...
+  </div>
+
+  <table border="0" cellpadding="10" cellspacing="10" align="center" width="100%" style="background-color: #f6f7f9;">
+    <tr>
+      <td>
+        <table align="center" border="0" cellpadding="0" cellspacing="0" width="600"
+          style="font-family: Helvetica,Arial; border-collapse: collapse; background-color: #ffffff; border-radius: 4px; border: solid 1px #d8d8d8;">
+          <tr>
+            <td bgcolor="#f3f5f6"
+              background="https://raw.githubusercontent.com/fylein/fyle-netsuite-api/b1ebe6c1db08c4e85cc34244f8bf5a1b3f1219ea/apps/workspaces/templates/images/background.png"
```
Comment (Sravanksk) [inline apps/workspaces/templates/mail_template.html:25]: these images are pointing to netsuite-api, please fix

File: apps/workspaces/templates/mail_template.html [hunk 7]
```diff
@@ -0,0 +1,206 @@
...
+              </table>
+            </td>
+          </tr>
+          <tr>
+            <td style="padding: 35px 25px 0px 15px;">
+              <table width="100%" cellspacing="10">
+                <tr align="center">
+                  <td>
+                    <img style="height: 150px;"
+                      src="https://raw.githubusercontent.com/fylein/fyle-netsuite-api/e05abf142208368b88a530f3d8bb25456fef4d26/apps/workspaces/templates/images/icon.png">
```
Comment (Sravanksk) [inline apps/workspaces/templates/mail_template.html:52]: same

File: apps/workspaces/templates/mail_template.html [hunk 8]
```diff
@@ -0,0 +1,206 @@
...
+          style="font-family: Helvetica,Arial; border-collapse: collapse; background-color: #ffffff; border-radius: 4px; border: solid 1px #d8d8d8;">
+          <tr>
+            <td bgcolor="#f3f5f6"
+              background="https://raw.githubusercontent.com/fylein/fyle-netsuite-api/b1ebe6c1db08c4e85cc34244f8bf5a1b3f1219ea/apps/workspaces/templates/images/background.png"
+              style="padding: 10px 25px 10px 25px;" width="600" height="50">
+
+              <table width="35%">
+                <tr>
+                  <td>
+                    <img src="https://sage-intacct.fyleapps.tech/assets/images/pngs/fyle-logo.png" alt="Fyle" title="Fyle"
```
Comment (Sravanksk) [inline apps/workspaces/templates/mail_template.html:31]: same, this is pointing to intacct staging assets

File: apps/workspaces/templates/mail_template.html [hunk 9]
```diff
@@ -0,0 +1,206 @@
...
+                      style="display:block" height="23" width="67" />
+                  </td>
+                  <td align="right">
+                    <span style="padding-left:8px">+</span>
+                  </td>
+                  <td align="right">
+                    <img
+                      src="https://sage-intacct.fyleapps.tech/assets/images/pngs/sage-intacct-logo.png"
```
Comment (Sravanksk) [inline apps/workspaces/templates/mail_template.html:39]: same

File: apps/workspaces/templates/mail_template.html [hunk 10]
```diff
@@ -0,0 +1,206 @@
...
+                      style="font-family: Aktiv Grotesk; padding-left: 10px;  background:#F5F5F5; border-radius: 8px;"
+                      width="100%" cellspacing="10">
+                      <tr style="font-style: normal;
+                          font-weight: 500;
+                          font-size: 14px;
+                          line-height: 15px;
+                          letter-spacing: 0.02em;
+                          color: #5A5D72;">
+                        <td>FYLE ORGANIZATION</td>
+                        <td>SAGE INTACCT ENTITY</td>
```
Comment (Sravanksk) [inline apps/workspaces/templates/mail_template.html:73]: remove sage-intacct references everywhere

File: apps/workspaces/templates/mail_template.html [hunk 11]
```diff
@@ -0,0 +1,206 @@
...
+                      </tr>
+                      <tr style="font-family: Aktiv Grotesk;
+                          font-style: normal;
+                          font-weight: 500;
+                          font-size: 14px;
+                          line-height: 20px;   
+                          color: #161528;">
+                        <td>{{ fyle_company }}</td>
+                        <td>{{ intacct_company }}</td>
```
Comment (Sravanksk) [inline apps/workspaces/templates/mail_template.html:82]: same

File: apps/workspaces/templates/mail_template.html [hunk 12]
```diff
@@ -0,0 +1,206 @@
...
+                      <tr align="left">
+                        <td style="font-size: 15px; line-height: 18px; color: #2C304E;">
+                          Dear {{ name }},
+                        </td>
+                      </tr>
+                      <tr>
+                        <td style="font-size: 15px; line-height: 18px; padding-top: 15px; color: #2C304E;">
+                          <span style="font-weight: bold">{{ errors }} expense groups</span> failed to export from your
+                          Fyle org
+                          to Sage Intacct Entity
```
Comment (Sravanksk) [inline apps/workspaces/templates/mail_template.html:99]: same

File: apps/workspaces/templates/mail_template.html [hunk 13]
```diff
@@ -0,0 +1,206 @@
...
+                      </tr>
+                      <tr>
+                        <td
+                          style="font-size: 15px; font-weight: bold; padding-top: 15px; line-height: 18px; color: #2C304E;">
+                          Export time - {{ export_time }}
+                        </td>
+                      </tr>
+                      <tr>
+                        <td style="font-size: 15px; line-height: 18px; padding-top: 15px; color: #2C304E;">
+                          Typically, expenses fail to export because of missing mapping or Sage Intacct authorization issues
```
Comment (Sravanksk) [inline apps/workspaces/templates/mail_template.html:110]: same

File: apps/workspaces/models.py (hunk 1 above)
Comment (DhaaraniCIT) [reply apps/workspaces/models.py:69]: related name has reference in advanced setting api

File: apps/workspaces/tasks.py (hunk 5 above)
Comment (DhaaraniCIT) [reply apps/workspaces/tasks.py:213]: It is already in setting.py

File: apps/workspaces/tasks.py [hunk 14]
```diff
@@ -1,24 +1,62 @@
-from datetime import datetime
+import time
```
Comment (NileshPant1999) [inline apps/workspaces/tasks.py:1]: no reference

File: apps/workspaces/tasks.py [hunk 15]
```diff
@@ -1,24 +1,62 @@
...
-def schedule_sync(workspace_id: int, schedule_enabled: bool, hours: int):
+def schedule_email_notification(workspace_id: int, schedule_enabled: bool, hours: int):
+    if schedule_enabled:
+        schedule, _ = Schedule.objects.update_or_create(
+            func='apps.workspaces.tasks.run_email_notification',
+            args='{}'.format(workspace_id),
+            defaults={
+                'schedule_type': Schedule.MINUTES,
+                'minutes': 5,
+                'next_run': datetime.now() + timedelta(minutes=1)
```
Comment (NileshPant1999) [inline apps/workspaces/tasks.py:31]: remember to change this later after testing

File: fyle_qbo_api/settings.py [hunk 16]
```diff
@@ -262,6 +263,16 @@
 QBO_ENVIRONMENT = os.environ.get('QBO_ENVIRONMENT')
 
 # Cache Settings
+SENDGRID_SANDBOX_MODE_IN_DEBUG=False
+
+# echo to stdout or any other file-like object that is passed to the backend via the stream kwarg.
+SENDGRID_ECHO_TO_STDOUT=True
```
Comment (NileshPant1999) [inline fyle_qbo_api/settings.py:269]: I need to test one thing here, set it to false in the local system and see if the email body still shows up in the terminal

File: requirements.txt [hunk 17]
```diff
@@ -41,3 +41,4 @@ pytest-cov==2.12.1
 pytest-django==4.4.0
 pytest-mock==3.6.1
 wrapt==1.12.1
+django-sendgrid-v5==1.2.0
```
Comment (NileshPant1999) [inline requirements.txt:44]: Can we place it alphabetically ??

File: apps/workspaces/views.py [hunk 18]
```diff
@@ -421,16 +422,21 @@ def post(self, request, **kwargs):
         """
         Post Settings
         """
-        schedule_enabled = request.data.get('schedule_enabled')
```
Comment (NileshPant1999) [inline apps/workspaces/views.py:424]: why are we changing this, will be better if we get them in the previous format, the camel case is not a good practice in python

File: apps/workspaces/templates/mail_template.html [hunk 19]
```diff
@@ -0,0 +1,256 @@
...
+              </table>
+            </td>
+          </tr>
+        </table>
+      </td>
+    </tr>
+  </table>
+</body>
+
+</html>
```
Comment (NileshPant1999) [inline apps/workspaces/templates/mail_template.html:256]: new line

File: fyle_qbo_api/settings.py (hunk 16 above)
Comment (NileshPant1999) [reply fyle_qbo_api/settings.py:269]: @DhaaraniCIT did you test this ??

Comment (DhaaraniCIT) [reply fyle_qbo_api/settings.py:269]: yes, if it is false it does'n show the email html code in terminal

File: apps/workspaces/models.py [hunk 20]
```diff
@@ -65,11 +66,14 @@ class WorkspaceSchedule(models.Model):
     Workspace Schedule
     """
     id = models.AutoField(primary_key=True, help_text='Unique Id to identify a schedule')
-    workspace = models.OneToOneField(Workspace, on_delete=models.PROTECT, help_text='Reference to Workspace model',
-                                     related_name='workspace_schedules')
+    workspace = models.OneToOneField(Workspace, on_delete=models.PROTECT, help_text='Reference to Workspace model',
```
Comment (Sravanksk) [inline apps/workspaces/models.py:69]: remove extra space, this line change is not needed

File: apps/workspaces/tasks.py [hunk 21]
```diff
@@ -1,24 +1,62 @@
...
-from apps.fyle.models import ExpenseGroup
 from apps.fyle.tasks import async_create_expense_groups
+from apps.fyle.models import Expense, ExpenseGroup
+from apps.fyle.serializers import ExpenseSerializer, ExpenseGroupSerializer
 from apps.quickbooks_online.tasks import schedule_bills_creation, schedule_cheques_creation, \
     schedule_journal_entry_creation, schedule_credit_card_purchase_creation, schedule_qbo_expense_creation
 from apps.tasks.models import TaskLog
-from apps.workspaces.models import WorkspaceSchedule, WorkspaceGeneralSettings, LastExportDetail
+from apps.workspaces.models import User, Workspace, WorkspaceSchedule, WorkspaceGeneralSettings, LastExportDetail, QBOCredential
+from fyle_accounting_mappings.models import ExpenseAttribute
```
Comment (Sravanksk) [inline apps/workspaces/tasks.py:19]: place this after line 10 ( module imports should be above app imports )

File: apps/workspaces/models.py (hunk 20 above)
Comment (DhaaraniCIT) [reply apps/workspaces/models.py:69]: this line is required

File: apps/workspaces/tasks.py [hunk 22]
```diff
@@ -136,8 +174,99 @@ def export_to_qbo(workspace_id, export_mode=None):
...
             )
-
     if is_expenses_exported:
         last_export_detail.last_exported_at = last_exported_at
         last_export_detail.export_mode = export_mode or 'MANUAL'
         last_export_detail.save()
+
+def run_email_notification(workspace_id):
+    expense_data = []
+    expence_html = ''
```
Comment (Sravanksk) [inline apps/workspaces/tasks.py:184]: fix typo expence_html -> `expense_html`

File: apps/workspaces/tasks.py [hunk 23]
```diff
@@ -136,8 +174,99 @@ def export_to_qbo(workspace_id, export_mode=None):
...
+    ws_schedule, _ = WorkspaceSchedule.objects.get_or_create(
+        workspace_id=workspace_id
+    )
+
+    task_logs = TaskLog.objects.filter(
+        ~Q(type__in=['CREATING_REIMBURSEMENT', 'FETCHING_EXPENSES', 'CREATING_AP_PAYMENT']),
```
Comment (Sravanksk) [inline apps/workspaces/tasks.py:190]: This is not fixed yet, please check different types of tasks in qbo_staging db and add filters accordingly

File: apps/workspaces/tasks.py [hunk 24]
```diff
@@ -136,8 +174,99 @@ def export_to_qbo(workspace_id, export_mode=None):
...
+        workspace_id=workspace_id,
+        status='FAILED'
+    )
+    workspace = Workspace.objects.get(id=workspace_id)
+    admin_data = WorkspaceSchedule.objects.get(workspace_id=workspace_id)
+    qbo = QBOCredential.objects.get(workspace=workspace)
+    errors = Error.objects.filter(workspace_id=workspace_id)
+    count = 0
+    for error in errors:
+        if error.type == 'EMPLOYEE_MAPPING' or error.type == 'CATEGORY_MAPPING' and count < 5:
```
Comment (Sravanksk) [inline apps/workspaces/tasks.py:200]: why are we checking for count < 5? Can you explain or add a comment to the code?

File: apps/workspaces/tasks.py [hunk 25]
```diff
@@ -136,8 +174,99 @@ def export_to_qbo(workspace_id, export_mode=None):
...
+            org_type = error.type.split('_')
+            error_type = org_type[0][0]+org_type[0][1:].lower()+' '+org_type[1][0]+org_type[1][1:].lower()
```
Comment (Sravanksk) [inline apps/workspaces/tasks.py:202]: fix indentation

File: apps/workspaces/tasks.py [hunk 26]
```diff
@@ -136,8 +174,99 @@ def export_to_qbo(workspace_id, export_mode=None):
...
+            expense_data.append(error_type)
+            html = '''<tr>
+                        <td> Mapping Error</td>
+                        <td>''' + error_type + '''</td>
+                        <td>''' + error.error_title + '''</td>
+                    </tr>'''
+            expence_html = expence_html + html
```
Comment (Sravanksk) [inline apps/workspaces/tasks.py:209]: expence_html = expence_html + html -> `expense_html = '{0} {1}'.format(expense_html, html)`

File: apps/workspaces/tasks.py [hunk 27]
```diff
@@ -136,8 +174,99 @@ def export_to_qbo(workspace_id, export_mode=None):
...
+        elif error.type == 'TAX_MAPPING' and count < 5:
+            expense_data.append('Tax Mapping')
+            org_type = error.type.split('_')
+            error_type = org_type[0][0]+org_type[0][1:].lower()+' '+org_type[1][0]+org_type[1][1:].lower()+' '+org_type[2][0]+org_type[2][1:].lower()
```
Comment (Sravanksk) [inline apps/workspaces/tasks.py:213]: fix indentation and also try to use string formatting everywhere ( `.format()` )

File: apps/workspaces/tasks.py [hunk 28]
```diff
@@ -136,8 +174,99 @@ def export_to_qbo(workspace_id, export_mode=None):
...
+            html = '''<tr>
+                        <td> Tax Mapping Error</td>
+                        <td>''' + error_type + '''</td>
+                        <td>''' + error.error_title + '''</td>
+                    </tr>'''
+            expence_html = expence_html + html
```
Comment (Sravanksk) [inline apps/workspaces/tasks.py:219]: same

File: apps/workspaces/tasks.py [hunk 29]
```diff
@@ -136,8 +174,99 @@ def export_to_qbo(workspace_id, export_mode=None):
...
+        elif error.type == 'QBO_ERROR' and count < 5:
+            expense_data.append('Quickbooks errors')
+            org_type = error.type.split('_')
+            error_type = org_type[0][0]+org_type[0][1:].lower()+' '+org_type[1][0]+org_type[1][1:].lower()
```
Comment (Sravanksk) [inline apps/workspaces/tasks.py:223]: same

File: apps/workspaces/tasks.py [hunk 30]
```diff
@@ -136,8 +174,99 @@ def export_to_qbo(workspace_id, export_mode=None):
...
+            html = '''<tr>
+                        <td> Quickbooks Errors </td>
+                        <td>''' + error_type + '''</td>
+                        <td>''' + error.error_title + '''</td>
+                    </tr>'''
+            expence_html = expence_html + html
```
Comment (Sravanksk) [inline apps/workspaces/tasks.py:229]: same

File: apps/workspaces/tasks.py [hunk 31]
```diff
@@ -136,8 +174,99 @@ def export_to_qbo(workspace_id, export_mode=None):
...
+                context = {
+                    'name': admin_name,
+                    'errors': len(task_logs),
+                    'fyle_company': workspace.name,
+                    'qbo_company': qbo.company_name,
+                    'workspace_id': workspace_id,
+                    'export_time': workspace.last_synced_at.strftime("%d %b %Y | %H:%M"),
+                    'year': date.today().year,
+                    'app_url': "{0}/workspaces/main/dashboard".format(settings.FYLE_APP_URL),
+                    'task_logs': mark_safe(expence_html),
```
Comment (Sravanksk) [inline apps/workspaces/tasks.py:256]: fix typo expence_html

File: apps/workspaces/tasks.py (hunk 24 above)
Comment (DhaaraniCIT) [reply apps/workspaces/tasks.py:200]: no of row in email is limited to 5, that why i have stated as 5

File: apps/workspaces/views.py [hunk 32]
```diff
@@ -548,3 +554,31 @@ def get(self, request, *args, **kwargs):
                 },
                 status=status.HTTP_400_BAD_REQUEST
             )
```
Comment (Sravanksk) [inline apps/workspaces/views.py:556]: add two line breaks after every class

File: apps/workspaces/tasks.py (hunk 25 above)
Comment (DhaaraniCIT) [reply apps/workspaces/tasks.py:202]: fix indentation means how you are meaning

**Review by Sravanksk — APPROVED** [review]

File: apps/workspaces/tasks.py [hunk 33]
```diff
@@ -136,8 +174,99 @@ def export_to_qbo(workspace_id, export_mode=None):
...
-
     if is_expenses_exported:
         last_export_detail.last_exported_at = last_exported_at
         last_export_detail.export_mode = export_mode or 'MANUAL'
         last_export_detail.save()
+
+def run_email_notification(workspace_id):
+    expense_data = []
+    expense_html = ''
+    ws_schedule, _ = WorkspaceSchedule.objects.get_or_create(
```
Comment (ashwin1111) [inline apps/workspaces/tasks.py:185]: why is this `get_or_create()` and not `.get()`? This function will be called by the worker only when the setting is enabled no?

File: apps/workspaces/tasks.py [hunk 34]
```diff
@@ -136,8 +174,99 @@ def export_to_qbo(workspace_id, export_mode=None):
...
+        workspace_id=workspace_id
+    )
+
+    task_logs = TaskLog.objects.filter(
+        ~Q(type__in=['CREATING_BILL_PAYMENT', 'FETCHING_EXPENSES']),
+        workspace_id=workspace_id,
+        status='FAILED'
+    )
+    workspace = Workspace.objects.get(id=workspace_id)
+    admin_data = WorkspaceSchedule.objects.get(workspace_id=workspace_id)
```
Comment (ashwin1111) [inline apps/workspaces/tasks.py:195]: We already have access to this in `ws_schedule`

File: apps/workspaces/tasks.py [hunk 35]
```diff
@@ -136,8 +174,99 @@ def export_to_qbo(workspace_id, export_mode=None):
...
+    qbo = QBOCredential.objects.get(workspace=workspace)
+    errors = Error.objects.filter(workspace_id=workspace_id)
```
Comment (ashwin1111) [inline apps/workspaces/tasks.py:197]: We should add `is_resolved=false` as well, else we might end up sending emails for resolved ones as well

File: apps/workspaces/tasks.py [hunk 36]
```diff
@@ -136,8 +174,99 @@ def export_to_qbo(workspace_id, export_mode=None):
...
+                        <td>''' + error_type + '''</td>
+                        <td>''' + error.error_title + '''</td>
+                    </tr>'''
+            expense_html = '{0} {1}'.format(expense_html, html)
+        else:
+            break
+        count = count+1
+    expense_data = set(expense_data)
+    expense_data = ','.join([str(data) for data in expense_data])
+    if ws_schedule.enabled:
```
Comment (ashwin1111) [inline apps/workspaces/tasks.py:235]: Why not add this filter directly in the above query? If it's false, all the above operations doesn't add any value.

File: apps/workspaces/tasks.py [hunk 37]
```diff
@@ -136,8 +174,99 @@ def export_to_qbo(workspace_id, export_mode=None):
...
+                    if data['email'] == admin_email:
+                        admin_name = data['name']
+
+            if task_logs and (ws_schedule.error_count is None or len(task_logs) > ws_schedule.error_count):
+                context = {
+                    'name': admin_name,
+                    'errors': len(task_logs),
+                    'fyle_company': workspace.name,
+                    'qbo_company': qbo.company_name,
+                    'workspace_id': workspace_id,
```
Comment (ashwin1111) [inline apps/workspaces/tasks.py:252]: where is this used?

File: apps/workspaces/tasks.py [hunk 38]
```diff
@@ -136,8 +174,99 @@ def export_to_qbo(workspace_id, export_mode=None):
...
+                    'export_time': workspace.last_synced_at.strftime("%d %b %Y | %H:%M"),
+                    'year': date.today().year,
+                    'app_url': "{0}/workspaces/main/dashboard".format(settings.FYLE_APP_URL),
```
Comment (ashwin1111) [inline apps/workspaces/tasks.py:255]: this needs to be added to https://github.com/fylein/fyle-deploy

File: fyle_qbo_api/settings.py [hunk 39]
```diff
@@ -262,6 +263,16 @@
 QBO_ENVIRONMENT = os.environ.get('QBO_ENVIRONMENT')
 
 # Cache Settings
+SENDGRID_SANDBOX_MODE_IN_DEBUG=False
+
+# echo to stdout or any other file-like object that is passed to the backend via the stream kwarg.
+SENDGRID_ECHO_TO_STDOUT=False
+SI_SENDER_ID = os.environ.get('SI_SENDER_ID')
```
Comment (ashwin1111) [inline fyle_qbo_api/settings.py:270]: where is this used?

File: fyle_qbo_api/settings.py [hunk 40]
```diff
@@ -262,6 +263,16 @@
...
+SI_SENDER_PASSWORD = os.environ.get('SI_SENDER_PASSWORD')
```
Comment (ashwin1111) [inline fyle_qbo_api/settings.py:271]: where is this used?

**Review by ashwin1111 — CHANGES_REQUESTED** [review]

File: fyle_qbo_api/settings.py (hunk 39 above)
Comment (NileshPant1999) [reply fyle_qbo_api/settings.py:270]: @ashwin1111 this is required by sendgrid

File: fyle_qbo_api/settings.py (hunk 40 above)
Comment (NileshPant1999) [reply fyle_qbo_api/settings.py:271]: same required by sendgrid

File: apps/workspaces/templates/mail_template.html (hunk 19 above)
Comment (NileshPant1999) [reply apps/workspaces/templates/mail_template.html:288]: @DhaaraniCIT why did you marked it as resolved ??

File: fyle_qbo_api/settings.py (hunk 39 above)
Comment (ashwin1111) [reply fyle_qbo_api/settings.py:270]: I see just send grid key and email used.

File: apps/workspaces/tasks.py (hunk 35 above)
Comment (NileshPant1999) [reply apps/workspaces/tasks.py:197]: @ashwin1111 we are only sending email when len(task_logs) > error_count, so even if the customer fixed some of the mappings we will not send the email again as the len(task_logs) then will be less then error count. can explain tomorrow if not clear.

Comment (ashwin1111) [reply apps/workspaces/tasks.py:197]: this returns all historical data which is not at all needed. assume a large org have few hundred errors. we are only concerned about non resolved errors

File: fyle_qbo_api/settings.py (hunk 39 above)
Comment (NileshPant1999) [reply fyle_qbo_api/settings.py:270]: this is in their document, so i added that, @DhaaraniCIT can you please try to remove it and see if it fails, may be this is used internally

Comment (ashwin1111) [reply fyle_qbo_api/settings.py:270]: this is Intacct sender ID and pwd saaar :D

Comment (ashwin1111) [reply fyle_qbo_api/settings.py:270]: where did you find it from the docs :P

Comment (NileshPant1999) [reply fyle_qbo_api/settings.py:270]: ohh i thought it is sendgrid ID 🤣

Comment (NileshPant1999) [reply fyle_qbo_api/settings.py:270]: My bad @DhaaraniCIT please remove this

Comment (DhaaraniCIT) [reply fyle_qbo_api/settings.py:270]: @NileshPant1999 and @ashwin1111 are you talking about this "SENDGRID_ECHO_TO_STDOUT" or "sendgrid ID"

File: apps/workspaces/tasks.py [hunk 41]
```diff
@@ -136,8 +174,81 @@ def export_to_qbo(workspace_id, export_mode=None):
...
+        expense_data = list(expense_data)
+        expense_data.append(error_type)
+        html_data = '''<td>''' + error_type + '''</td>
+                        <td>''' + error.error_title + '''</td>
+                    </tr>'''
+        html = '{0} {1}'.format(html,html_data)
+        expense_html = '{0} {1}'.format(expense_html, html)
+    expense_data = set(expense_data)
+    expense_data = ','.join([str(data) for data in expense_data])
+    if ws_schedule.enabled:
```
Comment (ashwin1111) [inline apps/workspaces/tasks.py:218]: Why is this still needed?

File: apps/workspaces/tasks.py [hunk 42]
```diff
@@ -136,8 +174,81 @@ def export_to_qbo(workspace_id, export_mode=None):
...
+            html = '''<tr>
+                        <td> Mapping Error </td>'''
+        elif error.type == 'TAX_MAPPING':
+            html = '''<tr>
+                        <td> Tax Mapping Error </td>'''
+        elif error.type == 'QBO_ERROR' and count < 5:
+                html = '''<tr>
+                            <td> Quickbooks Error </td>'''
+        org_type = error.type.split('_')
+        error_type = '{0} {1}'.format(org_type[0][0]+org_type[0][1:].lower(), org_type[1][0]+org_type[1][1:].lower())
```
Comment (ashwin1111) [inline apps/workspaces/tasks.py:208]: instead of doing this, can do this, looks more neat and not hacky `error.type.lower().title().replace('_', ' ')`

**Review by ashwin1111 — CHANGES_REQUESTED** [review]

File: apps/workspaces/tasks.py (hunk 41 above)
Comment (DhaaraniCIT) [reply apps/workspaces/tasks.py:218]: is this commend for **"if ws_schedule.enabled:"** this line

**Review by ashwin1111 — APPROVED** [review]

**Pull request merged**